kubernetes/istio.go

```diff
@@ -324,49 +325,63 @@ func DestinationRuleHasMTLSEnabled(destinationRule *networking_v1.DestinationRul
+func ClusterNameFromIstiod(conf *config.Config, k8s ClientInterface) (string, error) {
+	// The "cluster_id" is set in an environment variable of the "istiod" deployment. Let's try to fetch it.
+	istiods, err := k8s.GetDeployments("", metav1.ListOptions{LabelSelector: "app=istiod"})
+	if err != nil {
+		return "", err
+	}
 
-		istiodDeployment = deployment
-	} else {
-		istiodDeployments, err := k8s.GetDeployments(conf.IstioNamespace, metav1.ListOptions{LabelSelector: "app=istiod"})
-		if err != nil {
-			return "", err
-		}
+	if len(istiods) == 0 {
+		return "", fmt.Errorf("no valid istiod deployment found in any namespaces for kiali home cluster name discovery. You may need to set kubernetes_config.cluster_name in the Kiali CR")
+	}
 
-		if len(istiodDeployments) == 0 {
-			return "", fmt.Errorf("istiod deployment not found in namespace [%s]", conf.IstioNamespace)
+	// Do our best to filter out any external control plane, because the CLUSTER_ID will be set to
+	// the dataplane cluster, not the hosting cluster (which seems weird, but that's how Istio does it).
+	validIstiods := sliceutil.Filter(istiods, func(istiod v1.Deployment) bool {
+		// a typical namespace name is external-istiod
+		if strings.Contains(strings.ToLower(istiod.Namespace), "external") {
+			log.Tracef("Ignoring likely external controlplane [%s] during Kiali home clustername discovery. Namespace=[%s]", istiod.Name, istiod.Namespace)
+			return false
+		}
+		// there is a decent chance there will be an env var with "EXTERNAL"
+		for _, container := range istiod.Spec.Template.Spec.Containers {
+			for _, v := range container.Env {
+				if strings.Contains(strings.ToLower(v.Name), "external") {
+					log.Tracef("Ignoring likely external controlplane [%s] during Kiali home clustername discovery. EnvVar=[%s]", istiod.Name, v.Name)
+					return false
+				}
+			}
 		}
+		return true
+	})
 
-		// Just take the first one since they should all have the same cluster id.
-		istiodDeployment = &istiodDeployments[0]
+	// If we filtered out all of the istiods, just go back to what we had. In this case it's probably a primary-remote deployment where
+	// the primary may still have an EXTERNAL_ISTIOD env var.
+	if len(validIstiods) == 0 {
+		validIstiods = istiods
+		log.Tracef("Restoring original istiods as all were filtered. This is likely a primary-remote deployment.")
 	}
 
-	istiodContainers := istiodDeployment.Spec.Template.Spec.Containers
-	if len(istiodContainers) == 0 {
-		return "", fmt.Errorf("istiod deployment [%s] has no containers", istiodDeployment.Name)
+	validIstiod := validIstiods[0]
+	containers := validIstiod.Spec.Template.Spec.Containers
+	if len(containers) == 0 {
+		return "", fmt.Errorf("unexpected, istiod deployment [%s] has no containers", validIstiod.Name)
 	}
 
 	clusterName := ""
-	for _, v := range istiodContainers[0].Env {
+	for _, v := range containers[0].Env {
 		if v.Name == "CLUSTER_ID" {
 			clusterName = v.Value
-			break
```

Comment:
Why remove `break`?

Reply:
ooh, good eye. lost it when I removed some debug logging. I'll fix it.